java/core/src/main/java/com/whylogs/core/ColumnProfile.java

```diff
@@ -0,0 +1,82 @@
+package com.whylogs.core;
+
+import com.whylogs.core.metrics.Metric;
+import com.whylogs.core.metrics.OperationResult;
+import com.whylogs.core.schemas.ColumnSchema;
+import com.whylogs.core.views.ColumnProfileView;
+import java.util.ArrayList;
+import java.util.HashMap;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+
+@Getter
+@ToString
+@EqualsAndHashCode
+public class ColumnProfile<T> {
+  // Required
+  private final String name;
+  private final ColumnSchema schema;
+  private final int cacheSize;
+
+  // Has Defaults
+  private HashMap<String, Metric> metrics;
+  private final SingleFieldProjector<T> projector;
+  private int successCount;
+  private int failureCount;
+
+  private ArrayList<T> cachedValues;
+
+  public ColumnProfile(String name, ColumnSchema schema, int cacheSize) {
+    this.name = name;
+    this.schema = schema;
+    this.cacheSize = cacheSize; // TODO: add logger for size of cache on column
+
+    // Defaulted
+    this.metrics = new HashMap<>();
+    this.projector = new SingleFieldProjector<T>(name);
+    this.successCount = 0;
+    this.failureCount = 0;
+    this.cachedValues = new ArrayList<>();
+  }
+
+  public void addMetric(Metric metric) {
+    if (this.metrics.containsKey(metric.getNamespace())) {
+      // TODO: Add logger with warning about replacement
+    }
+
+    this.metrics.put(metric.getNamespace(), metric);
+  }
+
+  // TODO: this only gets one not every part of the row. Should projector actually do it multiple?
+  public void track(HashMap<String, Object> row) {
+    T value = this.projector.apply(row);
+    this.cachedValues.add(value);
+
+    if (this.cachedValues.size() >= this.cacheSize) {
+      this.flush();
+    }
+  }
+
```

Comment:
Do we have flush on destruction, or do we leave cached values orphaned?

Reply:
Good question. I was looking into this, I guess Java doesn't have a deconstructor and finalize has been deprecated. I went ahead and implemented AutoClosable. That should do it.

Reply:
I thought through this more, and I don't think it's a problem to orphan values... Any operation on the view that cares flushes the cache. If you have cached values but don't do a flushing operation on the profile the orphaned values don't matter.